Mac M1 , Qemu2 Driver Not found

### What Happened?

I ran minikube using qemu2 driver but I am getting a PROVIDER_QEMU2_NOT_FOUND error

``` shell
minikube start --driver=qemu2 --memory=26.5g --disk-size=270g --cpus=8 \
 --addons="[dashboard,default-storageclass,ingress,ingress-dns,metrics-server,registry,storage-provisioner]" \
 --kubernetes-version=1.23.6 --wait=all --dns-proxy=false --host-dns-resolver=true \
 --insecure-registry="10.0.0.0/16" --insecure-registry="192.168.49.0/24"
😄  minikube v1.26.0-beta.0 on Darwin 12.0.1 (arm64)
✨  Using the qemu2 (experimental) driver based on user configuration

🤷  Exiting due to PROVIDER_QEMU2_NOT_FOUND: The 'qemu2' provider was not found: exec: "qemu-system-aarch64": executable file not found in $PATH
💡  Suggestion: Install qemu-system
📘  Documentation: https://minikube.sigs.k8s.io/docs/reference/drivers/qemu2/
```
So I installed qemu using homebrew and now I am getting the following error

```shell
minikube start --driver=qemu2 --memory=26.5g --disk-size=270g --cpus=8 \
 --addons="[dashboard,default-storageclass,ingress,ingress-dns,metrics-server,registry,storage-provisioner]" \
 --kubernetes-version=1.23.6 --wait=all --dns-proxy=false --host-dns-resolver=true \
 --insecure-registry="10.0.0.0/16" --insecure-registry="192.168.49.0/24"
😄  minikube v1.26.0-beta.0 on Darwin 12.0.1 (arm64)
✨  Using the qemu2 (experimental) driver based on user configuration

🤷  Exiting due to PROVIDER_QEMU2_NOT_FOUND: The 'qemu2' provider was not found: stat /opt/homebrew/Cellar/qemu/6.2.0_1/share/qemu/edk2-aarch64-code.fd: no such file or directory
💡  Suggestion: Install uefi firmware
📘  Documentation: https://minikube.sigs.k8s.io/docs/reference/drivers/qemu2/

```


### Attach the log file

```shell
Binary: Built with gc go1.18.1 for darwin/arm64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I0516 14:24:13.513114   28241 out.go:296] Setting OutFile to fd 1 ...
I0516 14:24:13.513240   28241 out.go:348] isatty.IsTerminal(1) = true
I0516 14:24:13.513242   28241 out.go:309] Setting ErrFile to fd 2...
I0516 14:24:13.513244   28241 out.go:348] isatty.IsTerminal(2) = true
I0516 14:24:13.513294   28241 root.go:322] Updating PATH: /Users/dineshsankuri/.minikube/bin
I0516 14:24:13.513506   28241 out.go:303] Setting JSON to false
I0516 14:24:13.535424   28241 start.go:115] hostinfo: {"hostname":"Emplifis-MacBook-Pro.local","uptime":15888,"bootTime":1652709565,"procs":468,"os":"darwin","platform":"darwin","platformFamily":"Standalone Workstation","platformVersion":"12.0.1","kernelVersion":"21.1.0","kernelArch":"arm64","virtualizationSystem":"","virtualizationRole":"","hostId":"283ed594-67ba-5a7a-afbd-6537a2c07502"}
W0516 14:24:13.535530   28241 start.go:123] gopshost.Virtualization returned error: not implemented yet
I0516 14:24:13.556486   28241 out.go:177] 😄  minikube v1.26.0-beta.0 on Darwin 12.0.1 (arm64)
I0516 14:24:13.595928   28241 notify.go:193] Checking for updates...
I0516 14:24:13.596059   28241 driver.go:358] Setting default libvirt URI to qemu:///system
I0516 14:24:13.616017   28241 out.go:177] ✨  Using the qemu2 (experimental) driver based on user configuration
I0516 14:24:13.652781   28241 start.go:284] selected driver: qemu2
I0516 14:24:13.652787   28241 start.go:806] validating driver "qemu2" against <nil>
I0516 14:24:13.652801   28241 start.go:817] status for qemu2: {Installed:false Healthy:false Running:false NeedsImprovement:false Error:stat /opt/homebrew/Cellar/qemu/6.2.0_1/share/qemu/edk2-aarch64-code.fd: no such file or directory Reason: Fix:Install uefi firmware Doc:https://minikube.sigs.k8s.io/docs/reference/drivers/qemu2/ Version:}
I0516 14:24:13.671828   28241 out.go:177]
W0516 14:24:13.691417   28241 out.go:239] 🤷  Exiting due to PROVIDER_QEMU2_NOT_FOUND: The 'qemu2' provider was not found: stat /opt/homebrew/Cellar/qemu/6.2.0_1/share/qemu/edk2-aarch64-code.fd: no such file or directory
W0516 14:24:13.691501   28241 out.go:239] 💡  Suggestion: Install uefi firmware
W0516 14:24:13.691528   28241 out.go:239] 📘  Documentation: https://minikube.sigs.k8s.io/docs/reference/drivers/qemu2/
I0516 14:24:13.728811   28241 out.go:177]
```


### Operating System

macOS (Default)

### Driver

N/A

Sorry about using the image. I updated the post.

In homebrew the latest version of qemu is 6.2.0  so I tried to build the current version 7.0.0 from source code but I am having issues

```shell
In file included from /Users/dineshsankuri/Software/qemu-7.0.0/include/qemu/cpuid.h:14:
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:14:2: error: this header is for x86 only
#error this header is for x86 only
 ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:286:5: error: invalid output constraint '=a' in asm
    __cpuid(__leaf, __eax, __ebx, __ecx, __edx);
    ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:251:11: note: expanded from macro '__cpuid'
        : "=a"(__eax), "=r" (__ebx), "=c"(__ecx), "=d"(__edx) \
          ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:301:5: error: invalid output constraint '=a' in asm
    __cpuid(__leaf, *__eax, *__ebx, *__ecx, *__edx);
    ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:251:11: note: expanded from macro '__cpuid'
        : "=a"(__eax), "=r" (__ebx), "=c"(__ecx), "=d"(__edx) \
          ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:315:5: error: invalid output constraint '=a' in asm
    __cpuid_count(__leaf, __subleaf, *__eax, *__ebx, *__ecx, *__edx);
    ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:258:11: note: expanded from macro '__cpuid_count'
        : "=a"(__eax), "=r" (__ebx), "=c"(__ecx), "=d"(__edx) \
          ^
In file included from ../tcg/tcg.c:431:
/Users/dineshsankuri/Software/qemu-7.0.0/tcg/i386/tcg-target.c.inc:1041:30: warning: cast to 'const void *' from smaller integer type 'tcg_target_long' (aka 'int') [-Wint-to-void-pointer-cast]
    diff = tcg_pcrel_diff(s, (const void *)arg) - 7;
                             ^~~~~~~~~~~~~~~~~
/Users/dineshsankuri/Software/qemu-7.0.0/tcg/i386/tcg-target.c.inc:4088:9: error: invalid output constraint '=a' in asm
        __cpuid_count(7, 0, a, b7, c7, d);
        ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:258:11: note: expanded from macro '__cpuid_count'
        : "=a"(__eax), "=r" (__ebx), "=c"(__ecx), "=d"(__edx) \
          ^
In file included from ../tcg/tcg.c:431:
/Users/dineshsankuri/Software/qemu-7.0.0/tcg/i386/tcg-target.c.inc:4094:9: error: invalid output constraint '=a' in asm
        __cpuid(1, a, b, c, d);
        ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:251:11: note: expanded from macro '__cpuid'
        : "=a"(__eax), "=r" (__ebx), "=c"(__ecx), "=d"(__edx) \
          ^
In file included from ../tcg/tcg.c:431:
/Users/dineshsankuri/Software/qemu-7.0.0/tcg/i386/tcg-target.c.inc:4114:44: error: invalid output constraint '=a' in asm
            asm(".byte 0x0f, 0x01, 0xd0" : "=a" (xcrl), "=d" (xcrh) : "c" (0));
                                           ^
/Users/dineshsankuri/Software/qemu-7.0.0/tcg/i386/tcg-target.c.inc:4140:9: error: invalid output constraint '=a' in asm
        __cpuid(0x80000001, a, b, c, d);
        ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/cpuid.h:251:11: note: expanded from macro '__cpuid'
        : "=a"(__eax), "=r" (__ebx), "=c"(__ecx), "=d"(__edx) \
          ^
1 warning and 8 errors generated.
ninja: build stopped: subcommand failed.
make[1]: *** [run-ninja] Error 1
make: *** [all] Error 2
```

--- a/pkg/drivers/qemu/qemu.go
+++ b/pkg/drivers/qemu/qemu.go
@@ -337,10 +337,6 @@ func (d *Driver) Start() error {
 
 	if d.MachineType != "" {
 		machineType := d.MachineType
-		if runtime.GOOS == "darwin" {
-			// highmem=off needed, see https://patchwork.kernel.org/project/qemu-devel/patch/20201126215017.41156-9-agraf@csgraf.de/#23800615 for details
-			machineType += ",highmem=off"
-		}
 		startCmd = append(startCmd,
 			"-M", machineType,
 		)

--- a/pkg/drivers/qemu/qemu.go
+++ b/pkg/drivers/qemu/qemu.go
@@ -270,6 +270,10 @@ func parsePortRange(rawPortRange string) (int, int, error) {
 
 	portRange := strings.Split(rawPortRange, "-")
 
+	if len(portRange) < 2 {
+		return 0, 0, errors.New("Invalid port range, must be at least of length 2")
+	}
+
 	minPort, err := strconv.Atoi(portRange[0])
 	if err != nil {
 		return 0, 0, errors.Wrap(err, "Invalid port range")
@@ -378,7 +382,7 @@ func (d *Driver) Start() error {
 	// hardware acceleration is important, it increases performance by 10x
 	if runtime.GOOS == "darwin" {
 		startCmd = append(startCmd, "-accel", "hvf")
-	} else if runtime.GOOS == "linux" {
+	} else if _, err := os.Stat("/dev/kvm"); err == nil && runtime.GOOS == "linux" {
 		startCmd = append(startCmd, "-accel", "kvm")
 	}
 

--- a/pkg/drivers/qemu/qemu.go
+++ b/pkg/drivers/qemu/qemu.go
@@ -380,10 +380,10 @@ func (d *Driver) Start() error {
 	}
 
 	// hardware acceleration is important, it increases performance by 10x
+	// kvm acceleration doesn't currently work for linux, it's incompatible with our chosen CPU
+	// once that's fixed we should add a branch for linux
 	if runtime.GOOS == "darwin" {
 		startCmd = append(startCmd, "-accel", "hvf")
-	} else if runtime.GOOS == "linux" {
-		startCmd = append(startCmd, "-accel", "kvm")
 	}
 
 	startCmd = append(startCmd,
@@ -427,9 +427,10 @@ func (d *Driver) Start() error {
 
 	// other options
 	// "-enable-kvm" if its available
-	if _, err := os.Stat("/dev/kvm"); err == nil {
+	// TODO (#14171): re-enable this once kvm acceleration is fixed
+	/*if _, err := os.Stat("/dev/kvm"); err == nil {
 		startCmd = append(startCmd, "-enable-kvm")
-	}
+	}*/
 
 	if d.CloudConfigRoot != "" {
 		startCmd = append(startCmd,

--- a/pkg/drivers/qemu/qemu.go
+++ b/pkg/drivers/qemu/qemu.go
@@ -380,10 +380,10 @@ func (d *Driver) Start() error {
 	}
 
 	// hardware acceleration is important, it increases performance by 10x
-	// kvm acceleration doesn't currently work for linux, it's incompatible with our chosen CPU
-	// once that's fixed we should add a branch for linux
 	if runtime.GOOS == "darwin" {
 		startCmd = append(startCmd, "-accel", "hvf")
+	} else if runtime.GOOS == "linux" {
+		startCmd = append(startCmd, "-accel", "kvm")
 	}
 
 	startCmd = append(startCmd,
@@ -424,13 +424,6 @@ func (d *Driver) Start() error {
 	}
 
 	startCmd = append(startCmd, "-daemonize")
-
-	// other options
-	// "-enable-kvm" if its available
-	// TODO (#14171): re-enable this once kvm acceleration is fixed
-	/*if _, err := os.Stat("/dev/kvm"); err == nil {
-		startCmd = append(startCmd, "-enable-kvm")
-	}*/
 
 	if d.CloudConfigRoot != "" {
 		startCmd = append(startCmd,

--- a/pkg/drivers/qemu/qemu.go
+++ b/pkg/drivers/qemu/qemu.go
@@ -339,7 +339,7 @@ func (d *Driver) Start() error {
 		machineType := d.MachineType
 		if runtime.GOOS == "darwin" {
 			// highmem=off needed, see https://patchwork.kernel.org/project/qemu-devel/patch/20201126215017.41156-9-agraf@csgraf.de/#23800615 for details
-			machineType += ",accel=hvf,highmem=off"
+			machineType += ",highmem=off"
 		}
 		startCmd = append(startCmd,
 			"-M", machineType,
@@ -377,6 +377,13 @@ func (d *Driver) Start() error {
 				"-display", "none",
 			)
 		}
+	}
+
+	// hardware acceleration is important, it increases performance by 10x
+	if runtime.GOOS == "darwin" {
+		startCmd = append(startCmd, "-accel", "hvf")
+	} else if runtime.GOOS == "linux" {
+		startCmd = append(startCmd, "-accel", "kvm")
 	}
 
 	startCmd = append(startCmd,

--- a/pkg/drivers/qemu/qemu.go
+++ b/pkg/drivers/qemu/qemu.go
@@ -38,6 +38,7 @@ import (
 	"github.com/docker/machine/libmachine/mcnutils"
 	"github.com/docker/machine/libmachine/ssh"
 	"github.com/docker/machine/libmachine/state"
+	"github.com/pkg/errors"
 
 	pkgdrivers "k8s.io/minikube/pkg/drivers"
 )
@@ -79,11 +80,9 @@ type Driver struct {
 	DiskPath         string
 	CacheMode        string
 	IOMode           string
-	//	conn             *libvirt.Connect
-	//	VM               *libvirt.Domain
-	UserDataFile    string
-	CloudConfigRoot string
-	LocalPorts      string
+	UserDataFile     string
+	CloudConfigRoot  string
+	LocalPorts       string
 }
 
 func (d *Driver) GetMachineName() string {
@@ -273,19 +272,19 @@ func parsePortRange(rawPortRange string) (int, int, error) {
 
 	minPort, err := strconv.Atoi(portRange[0])
 	if err != nil {
-		return 0, 0, fmt.Errorf("invalid port range")
+		return 0, 0, errors.Wrap(err, "Invalid port range")
 	}
 	maxPort, err := strconv.Atoi(portRange[1])
 	if err != nil {
-		return 0, 0, fmt.Errorf("invalid port range")
+		return 0, 0, errors.Wrap(err, "Invalid port range")
 	}
 
 	if maxPort < minPort {
-		return 0, 0, fmt.Errorf("invalid port range")
+		return 0, 0, errors.New("Invalid port range")
 	}
 
 	if maxPort-minPort < 2 {
-		return 0, 0, fmt.Errorf("port range must be minimum 2 ports")
+		return 0, 0, errors.New("Port range must be minimum 2 ports")
 	}
 
 	return minPort, maxPort, nil

--- a/pkg/minikube/registry/drvs/qemu2/qemu2.go
+++ b/pkg/minikube/registry/drvs/qemu2/qemu2.go
@@ -68,7 +68,6 @@ func qemuFirmwarePath() (string, error) {
 	arch := runtime.GOARCH
 	// For macOS, find the correct brew installation path for qemu firmware
 	if runtime.GOOS == "darwin" {
-		//return "/opt/homebrew/Cellar/qemu/6.2.0_1/share/qemu/edk2-aarch64-code.fd", nil
 		p := "/usr/local/Cellar/qemu"
 		fw := "share/qemu/edk2-x86_64-code.fd"
 		if arch == "arm64" {
@@ -86,6 +85,7 @@ func qemuFirmwarePath() (string, error) {
 			}
 		}
 	}
+
 	switch arch {
 	case "amd64":
 		return "/usr/share/OVMF/OVMF_CODE.fd", nil

--- a/pkg/minikube/registry/drvs/qemu2/qemu2.go
+++ b/pkg/minikube/registry/drvs/qemu2/qemu2.go
@@ -114,11 +114,12 @@ func configure(cc config.ClusterConfig, n config.Node) (interface{}, error) {
 		qemuMachine = "" // default
 		qemuCPU = ""     // default
 	case "arm64":
+		qemuMachine = "virt"
+		qemuCPU = "cortex-a72"
 		// highmem=off needed, see https://patchwork.kernel.org/project/qemu-devel/patch/20201126215017.41156-9-agraf@csgraf.de/#23800615 for details
 		if runtime.GOOS == "darwin" {
 			qemuMachine = "virt,highmem=off"
-			qemuCPU = "cortex-a72"
-		} else if runtime.GOOS == "linux" {
+		} else if _, err := os.Stat("/dev/kvm"); err == nil {
 			qemuMachine = "virt,gic-version=3"
 			qemuCPU = "host"
 		}

--- a/pkg/minikube/registry/drvs/qemu2/qemu2.go
+++ b/pkg/minikube/registry/drvs/qemu2/qemu2.go
@@ -68,11 +68,16 @@ func qemuFirmwarePath() (string, error) {
 	arch := runtime.GOARCH
 	// For macOS, find the correct brew installation path for qemu firmware
 	if runtime.GOOS == "darwin" {
-		p := "/usr/local/Cellar/qemu"
-		fw := "share/qemu/edk2-x86_64-code.fd"
-		if arch == "arm64" {
+		var p, fw string
+		switch arch {
+		case "amd64":
+			p = "/usr/local/Cellar/qemu"
+			fw = "share/qemu/edk2-x86_64-code.fd"
+		case "arm64":
 			p = "/opt/homebrew/Cellar/qemu"
 			fw = "share/qemu/edk2-aarch64-code.fd"
+		default:
+			return "", fmt.Errorf("unknown arch: %s", arch)
 		}
 
 		v, err := ioutil.ReadDir(p)
@@ -109,8 +114,13 @@ func configure(cc config.ClusterConfig, n config.Node) (interface{}, error) {
 		qemuMachine = "" // default
 		qemuCPU = ""     // default
 	case "arm64":
-		qemuMachine = "virt"
-		qemuCPU = "cortex-a72"
+		if runtime.GOOS == "darwin" {
+			qemuMachine = "virt"
+			qemuCPU = "cortex-a72"
+		} else if runtime.GOOS == "linux" {
+			qemuMachine = "gic-version=3"
+			qemuCPU = "host"
+		}
 	default:
 		return nil, fmt.Errorf("unknown arch: %s", runtime.GOARCH)
 	}

--- a/pkg/minikube/registry/drvs/qemu2/qemu2.go
+++ b/pkg/minikube/registry/drvs/qemu2/qemu2.go
@@ -18,8 +18,10 @@ package qemu2
 
 import (
 	"fmt"
+	"io/ioutil"
 	"os"
 	"os/exec"
+	"path"
 	"path/filepath"
 	"runtime"
 
@@ -64,17 +66,30 @@ func qemuSystemProgram() (string, error) {
 
 func qemuFirmwarePath() (string, error) {
 	arch := runtime.GOARCH
+	// For macOS, find the correct brew installation path for qemu firmware
+	if runtime.GOOS == "darwin" {
+		//return "/opt/homebrew/Cellar/qemu/6.2.0_1/share/qemu/edk2-aarch64-code.fd", nil
+		p := "/usr/local/Cellar/qemu"
+		fw := "share/qemu/edk2-x86_64-code.fd"
+		if arch == "arm64" {
+			p = "/opt/homebrew/Cellar/qemu"
+			fw = "share/qemu/edk2-aarch64-code.fd"
+		}
+
+		v, err := ioutil.ReadDir(p)
+		if err != nil {
+			return "", fmt.Errorf("lookup qemu: %v", err)
+		}
+		for _, version := range v {
+			if version.IsDir() {
+				return path.Join(p, version.Name(), fw), nil
+			}
+		}
+	}
 	switch arch {
 	case "amd64":
-		// on macOS, we assume qemu is installed via homebrew for simplicity
-		if runtime.GOOS == "darwin" {
-			return "/usr/local/Cellar/qemu/6.2.0_1/share/qemu/edk2-x86_64-code.fd", nil
-		}
 		return "/usr/share/OVMF/OVMF_CODE.fd", nil
 	case "arm64":
-		if runtime.GOOS == "darwin" {
-			return "/opt/homebrew/Cellar/qemu/6.2.0_1/share/qemu/edk2-aarch64-code.fd", nil
-		}
 		return "/usr/share/AAVMF/AAVMF_CODE.fd", nil
 	default:
 		return "", fmt.Errorf("unknown arch: %s", arch)

--- a/pkg/minikube/registry/drvs/qemu2/qemu2.go
+++ b/pkg/minikube/registry/drvs/qemu2/qemu2.go
@@ -114,11 +114,12 @@ func configure(cc config.ClusterConfig, n config.Node) (interface{}, error) {
 		qemuMachine = "" // default
 		qemuCPU = ""     // default
 	case "arm64":
+		// highmem=off needed, see https://patchwork.kernel.org/project/qemu-devel/patch/20201126215017.41156-9-agraf@csgraf.de/#23800615 for details
 		if runtime.GOOS == "darwin" {
-			qemuMachine = "virt"
+			qemuMachine = "virt,highmem=off"
 			qemuCPU = "cortex-a72"
 		} else if runtime.GOOS == "linux" {
-			qemuMachine = "gic-version=3"
+			qemuMachine = "virt,gic-version=3"
 			qemuCPU = "host"
 		}
 	default:

